src/packet/signature.js

```diff
@@ -660,33 +660,41 @@ class SignaturePacket {
   async verify(key, signatureType, data, date = new Date(), detached = false, config = defaultConfig) {
     const publicKeyAlgorithm = enums.write(enums.publicKey, this.publicKeyAlgorithm);
     const hashAlgorithm = enums.write(enums.hash, this.hashAlgorithm);
-
+    if (!this.issuerKeyID.equals(key.getKeyID())) {
+      throw new Error('Signature was not issued by the given public key');
+    }
     if (publicKeyAlgorithm !== enums.write(enums.publicKey, key.algorithm)) {
       throw new Error('Public key algorithm used to sign signature does not match issuer key algorithm.');
     }
 
-    let toHash;
-    let hash;
-    if (this.hashed) {
-      hash = await this.hashed;
-    } else {
-      toHash = this.toHash(signatureType, data, detached);
-      hash = await this.hash(signatureType, data, toHash);
-    }
-    hash = await stream.readToEnd(hash);
-    if (this.signedHashValue[0] !== hash[0] ||
-        this.signedHashValue[1] !== hash[1]) {
-      throw new Error('Message digest did not match');
-    }
+    const isMessageSignature = signatureType === enums.signature.binary || signatureType === enums.signature.text;
+    // Cryptographic validity is cached after one successful verification.
+    // However, for message signatures, we always re-verify, since the passed `data` can change
+    const skipVerify = this.verified && !isMessageSignature;
+    if (!skipVerify) {
+      let toHash;
+      let hash;
+      if (this.hashed) {
+        hash = await this.hashed;
+      } else {
+        toHash = this.toHash(signatureType, data, detached);
+        hash = await this.hash(signatureType, data, toHash);
+      }
+      hash = await stream.readToEnd(hash);
+      if (this.signedHashValue[0] !== hash[0] ||
+          this.signedHashValue[1] !== hash[1]) {
+        throw new Error('Invalid signature: Message digest did not match');
```

Comment:
Usually, this should be wrapped in something like "Could not find valid self-signature:" or "Message signature could not be verified:" (though some of those we still have to add at some point) and then this will be a bit redundant.